### PR TITLE
store: change max_open_files to u32

### PR DESCRIPTION
**Pull request by mina86**
Change type of max_open_files configurable setting from i32 to u32
so that we don’t have to deal with negative values.  For example,
we don’t need to worry about `max_open_files as u64` blowing up to
astronomical values.  Putting a negative value for the field is not
valid anyway.

### CODE REVIEW

**Comment by matklad** [comment]
apparently -1 is a valid valuer here? 

https://github.com/facebook/rocksdb/blob/d6e016be6dceae99adb2c068331c90d468661873/include/rocksdb/options.h#L548-L561

Do we want to support that?

**Comment by mina86** [comment]
Fair question. I was kind assuming we don’t. If we do than we need to update our logic for setting NOFILE limit.

File: core/store/src/db.rs [hunk 1]
```diff
@@ -481,7 +481,7 @@ fn rocksdb_options(store_config: &StoreConfig) -> Options {
     opts.create_missing_column_families(true);
     opts.create_if_missing(true);
     opts.set_use_fsync(false);
-    opts.set_max_open_files(store_config.max_open_files);
+    opts.set_max_open_files(store_config.max_open_files.try_into().unwrap_or(i32::MAX));
```
[the remark this answers is not in the document]
Comment (mina86) [reply core/store/src/db.rs:484]: I dunno. To me it doesn’t feel like an error to put a huge value here. The configuration says ‘open at most N’ files and we're here obeying that setting.

[the remark this answers is not in the document]
Comment (nagisa) [reply core/store/src/db.rs:484]: Right, I don't consider huge values an inherent problem (besides us not knowing how exactly rocksdb will behave when it ends up hitting the much lower limit imposed by the kernel.)

My concern here is largely in the code silently trying to be smarter than a human operator. 

To give an analogy, if I were in a position of turning knobs, and there was a knob that went from 0 to 11, and turning it towards 11 stopped making things more-11 somewhere midway, I would (ノ｀´)ノ ~┻━┻ pretty quick.

This `unwrap_or` is doing effectively the same thing – you can specify up to `u32::MAX` max_open_files without an error, but the config option would pretty much stop doing anything past `i32::MAX`.

I think it is better, code hygiene wise at least, to just say that `u16::MAX` “ought to be enough files for everybody” for now and have a fully working knob here.

**Comment by mina86** [comment]
We need to reach some consensus what to do here. I see three options:
1. Leave it at `i32` to allow -1 and fix NOLIMIT setting code which currently uses `as u32`.
2. Change it to `u32` prohibiting -1 and adding clipping to `i32::MAX`.
3. Change it to `u16` prohibiting -1 and values over 65k.

**Comment by matklad** [comment]
I'd say clipping (2) is best -- the actual thing is `c_int`, so using `u16` won't be pedantically correct. And I don't think we should leak details about what `c_int` is to our config layer, even if via errors. Given that the limit is approximate anyway (we add eyeballed headroom), the semantics "if you asked a whole lot of files, you get some unspecified whole lot of files" feels right to me. 

Broadly, I don't feel strongly about what we do internally, but I think it's worthwhile to encapsulate idiosyncratic rocks API inside and let the outside world deal with u32, instead of guessing what the limit of -92 files means :)

**Review by matklad — APPROVED** [review]

**Pull request merged**